### PR TITLE
Remove high cardanility metrics from otelhttp

### PR DESCRIPTION
**Pull request by MadVikingGod**
This does not upgrade to the latest semantic convention, that will be a different PR.

This removes the following attributes from the metrics that `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` generates:
- `net.sock.peer.addr`
- `net.sock.peer.port`
- `http.user_agent`
- `enduser.id`
- `http.client_ip`

This also only allows `http.request.method` to be known HTTP methods, or `_OTHER`.

Resolve https://github.com/open-telemetry/opentelemetry-go/issues/3765

### CODE REVIEW

**Review by hanyuancheung — APPROVED** [review]

File: instrumentation/github.com/gin-gonic/gin/otelgin/internal/semconvutil/httpconv.go [hunk 1]
```diff
@@ -349,21 +349,101 @@ func (c *httpConv) ServerRequest(server string, req *http.Request) []attribute.K
...
 		return c.HTTPMethodKey.String(http.MethodGet)
 	}
 	return c.HTTPMethodKey.String(method)
 }
 
+func (c *httpConv) methodMetric(method string) attribute.KeyValue {
+	method = strings.ToUpper(method)
+	if _, ok := allowedMethods[method]; !ok {
+		method = "_OTHER"
+	}
```
Comment (pellared) [inline instrumentation/github.com/gin-gonic/gin/otelgin/internal/semconvutil/httpconv.go:423]: This looks very Pythonish 😉 

How about using `switch` and constants from `http` package?

```go
	switch method {
	case http.MethodGet , http.MethodHead, http.MethodPost, ....
	default:
		method = "_OTHER"
	}
```

File: instrumentation/github.com/gin-gonic/gin/otelgin/internal/semconvutil/httpconv.go [hunk 2]
```diff
@@ -349,21 +349,101 @@ func (c *httpConv) ServerRequest(server string, req *http.Request) []attribute.K
...
+
 func (c *httpConv) method(method string) attribute.KeyValue {
 	if method == "" {
 		return c.HTTPMethodKey.String(http.MethodGet)
 	}
 	return c.HTTPMethodKey.String(method)
 }
 
+func (c *httpConv) methodMetric(method string) attribute.KeyValue {
+	method = strings.ToUpper(method)
```
[the remark this answers is not in the document]
Comment (MadVikingGod) [reply instrumentation/github.com/gin-gonic/gin/otelgin/internal/semconvutil/httpconv.go:420]: The semantic convention has a list of values allowed, and we are to convert equivalent levels into the canonical form.  If this were tracing, where we don't have cardinality concerns, we would also set `http.request.method_original`. https://github.com/open-telemetry/semantic-conventions/blob/main/model/http-common.yaml

File: instrumentation/github.com/emicklei/go-restful/otelrestful/internal/semconvutil/httpconv.go [hunk 3]
```diff
@@ -349,21 +349,91 @@ func (c *httpConv) ServerRequest(server string, req *http.Request) []attribute.K
...
+//
+// If the primary server name is not known, server should be an empty string.
+// The req Host will be used to determine the server instead.
+//
+// The following attributes are always returned: "http.method", "http.scheme",
+// "http.flavor", "http.target", "net.host.name". The following attributes are
+// returned if they related values are defined in req: "net.host.port",
+// "net.sock.peer.addr", "net.sock.peer.port", "http.user_agent", "enduser.id",
+// "http.client_ip".
+func (c *httpConv) ServerRequestMetrics(server string, req *http.Request) []attribute.KeyValue {
```
Comment (pellared) [inline instrumentation/github.com/emicklei/go-restful/otelrestful/internal/semconvutil/httpconv.go:374]: I think you should also create a `HTTPServerRequestMetrics` function (that will be used in `otelhttp`) and add tests.

File: instrumentation/github.com/gin-gonic/gin/otelgin/internal/semconvutil/httpconv.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pellared) [reply instrumentation/github.com/gin-gonic/gin/otelgin/internal/semconvutil/httpconv.go:420]: Thanks for clarification.

**Review by pellared — APPROVED** [review]

File: internal/shared/semconvutil/httpconv.go.tmpl [hunk 4]
```diff
@@ -349,21 +373,91 @@ func (c *httpConv) ServerRequest(server string, req *http.Request) []attribute.K
...
+// suffix.
+//
+// If the primary server name is not known, server should be an empty string.
+// The req Host will be used to determine the server instead.
+//
+// The following attributes are always returned: "http.method", "http.scheme",
+// "http.flavor", "http.target", "net.host.name". The following attributes are
+// returned if they related values are defined in req: "net.host.port",
+// "net.sock.peer.addr", "net.sock.peer.port", "http.user_agent", "enduser.id",
+// "http.client_ip".
```
Comment (pellared) [inline internal/shared/semconvutil/httpconv.go.tmpl:397]:
```suggestion
// The following attributes are always returned: "http.method", "http.scheme",
// "http.flavor", "net.host.name". The following attributes are
// returned if they related values are defined in req: "net.host.port".
```

File: internal/shared/semconvutil/httpconv.go.tmpl [hunk 5]
```diff
@@ -349,21 +373,91 @@ func (c *httpConv) ServerRequest(server string, req *http.Request) []attribute.K
 	return attrs
 }
 
+// ServerRequestMetrics returns Metric attributes for an HTTP request received
```
Comment (pellared) [inline internal/shared/semconvutil/httpconv.go.tmpl:376]:
```suggestion
// ServerRequestMetrics returns metric attributes for an HTTP request received
```

File: internal/shared/semconvutil/httpconv.go.tmpl [hunk 6]
```diff
@@ -83,6 +83,30 @@ func HTTPServerRequest(server string, req *http.Request) []attribute.KeyValue {
 	return hc.ServerRequest(server, req)
 }
 
+// HTTPServerRequestMetrics returns trace attributes for an HTTP request received by a
```
Comment (pellared) [inline internal/shared/semconvutil/httpconv.go.tmpl:86]:
```suggestion
// HTTPServerRequestMetrics returns metrics attributes for an HTTP request received by a
```

**Review by pellared — CHANGES_REQUESTED** [review]

**Review by pellared — APPROVED** [review]

**Comment by davendu** [comment]
Weird. Why @dmathieu 's [previous suggestion](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3980#pullrequestreview-1476833100) is not applied to this PR:

> I don't think this is the proper approach. It's adding some weird complexity, and whether an attribute has too much cardinality or not is very subjective.
> 
> The solution being discussed in [open-telemetry/opentelemetry-specification#3061](https://github.com/open-telemetry/opentelemetry-specification/issues/3061) seems much better.


And what's the community's suggestion towards similar attributes? My own instrumentation library also needs some `high cardanility` attributes, and not sure if I should remove them or filter through the view.

**Comment by dmathieu** [comment]
This PR solves from a potential memory leak issue, since the values from those attributes are coming from untrusted sources.
We should be able to bring back those attributes once something like https://github.com/open-telemetry/opentelemetry-go/pull/4457 is stable.

**Comment by FaranIdo** [comment]
@dmathieu @pellared @MadVikingGod  do you plan to add a similar solution to otelgrpc to avoid high cardinality there as well? (I think reported here: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3071)

**Pull request merged**